Source/GMCAbilitySystem/Public/Utility/GMASBoundQueue.h

```diff
@@ -0,0 +1,582 @@
+
+UENUM(BlueprintType)
+enum class EGMASBoundQueueOperationType : uint8
+{
+	None,
+	Add,
+	Remove,
+	Activate,
+	Cancel
+};
+
+USTRUCT(BlueprintType)
+struct FGMASBoundQueueOperationIdSet
+{
+	GENERATED_BODY()
+
+	UPROPERTY()
+	TArray<int> Ids = {};
+};
+
+USTRUCT(BlueprintType)
+struct GMCABILITYSYSTEM_API FGMASBoundQueueRPCHeader
+{
+	GENERATED_BODY()
+
+	UPROPERTY()
+	int32 OperationId { -1 };
+
+	UPROPERTY()
+	uint8 OperationTypeRaw { 0 };
+
+	EGMASBoundQueueOperationType GetOperationType() const
+	{
+		return static_cast<EGMASBoundQueueOperationType>(OperationTypeRaw);
+	}
+	
+	UPROPERTY()
+	FGameplayTag Tag { FGameplayTag::EmptyTag };
+
+	UPROPERTY()
+	FName ItemClassName { NAME_None };
+
+	UPROPERTY()
+	FGMASBoundQueueOperationIdSet PayloadIds {};
+
+	UPROPERTY()
+	FInstancedStruct InstancedPayload {};
+
+	UPROPERTY()
+	float RPCGracePeriodSeconds { 1.f };
+
+	UPROPERTY()
+	uint8 ExtraFlags { 0 };
+};
+
+template<typename C, typename T>
+struct GMCABILITYSYSTEM_API TGMASBoundQueueOperation
+{
+	FGMASBoundQueueRPCHeader Header {};
+	
+	// An actual class to be utilized with this, in case we need to instance it.
+	TSubclassOf<C> ItemClass { nullptr };
+
+	FInstancedStruct InstancedPayloadIds;
+	
+	// The struct payload for this item.
+	T Payload;	
+
+	// The instanced struct representation of this payload, used to actually
+	// bind for replication.
+	FInstancedStruct InstancedPayload;
+
+	EGMASBoundQueueOperationType GetOperationType() const
+	{
+		return Header.GetOperationType();
+	}
+
+	void SetOperationType(EGMASBoundQueueOperationType OperationType)
+	{
+		Header.OperationTypeRaw = static_cast<uint8>(OperationType);
+	}
+
+	int32 GetOperationId() const { return Header.OperationId; }
+
+	TArray<int32> GetPayloadIds() const { return Header.PayloadIds.Ids; }
+
+	FGameplayTag GetTag() const { return Header.Tag; }
+
+	bool GracePeriodExpired() const
+	{
+		return Header.RPCGracePeriodSeconds <= 0.f;
+	}
+
+	bool IsValid() const
+	{
+		return Header.OperationTypeRaw != 0 && (Header.ItemClassName != NAME_None || Header.Tag != FGameplayTag::EmptyTag || Header.PayloadIds.Ids.Num() > 0);
+	}
+
+	void Refresh(bool bDecodePayload = false)
+	{
+		if (bDecodePayload)
+		{
+			// Incoming from remote.
+			Payload = Header.InstancedPayload.Get<T>();
+			if (Header.PayloadIds.Ids.Num() == 0 && InstancedPayloadIds.IsValid())
+			{
+				Header.PayloadIds = InstancedPayloadIds.Get<FGMASBoundQueueOperationIdSet>();
+			}
+		}
+		else
+		{
+			// Outgoing
+			Header.InstancedPayload = FInstancedStruct::Make<T>(Payload);
+			InstancedPayloadIds = FInstancedStruct::Make<FGMASBoundQueueOperationIdSet>(Header.PayloadIds);
+		}
+
+		RefreshClass();
+	}
+
+	void RefreshClass()
+	{
+		if (Header.ItemClassName != NAME_None && !ItemClass)
+		{
+			// Get a handle to our class, for instancing purposes.
+			TSoftClassPtr<C> ClassPtr = TSoftClassPtr<C>(FSoftObjectPath(Header.ItemClassName.ToString()));
+			ItemClass = ClassPtr.LoadSynchronous();
```

Comment:
This can cause hitches as this flushes the async streaming queue. I would use the asset manager's `FStreamableManager::RequestSyncLoad()` as it instead puts the request at the top of the queue, potentially finishing faster.

Reply:
Reasonable change!